lib/pushgateway.js

```diff
@@ -75,7 +75,14 @@ function useGateway(method, job, groupings, callback) {
 	});
 
 	if (method !== 'DELETE') {
-		req.write(this.registry.metrics());
+		this.registry
+			.metrics()
+			.then(metrics => {
+				req.write(metrics);
+			})
+			.catch(e => {
+				callback(e);
+			});
 	}
 	req.end();
```

Comment:
In the current implementation, there is a subtle bug, `req.end()` might happen before `req.write(metrics)` is called because the async version of `registry.metrics()` is non-blocking.
I suggest making `useGateway()` async. This is the only easy way of keeping the existing flow semantics.
```javascript
async function useGateway(method, job, groupings, callback) {

        ...

	if (method !== 'DELETE') {
		try {
			const metrics = await this.registry.metrics();
			req.write(metrics);
		} catch (error) {
			callback(error);
		}
	}
	req.end();
}
```

Reply:
@danielkatz I don't really want to mix callbacks and `async/await`.

@zD98 danielkatz is correct about the bug though. Below is the fix I was thinking of:

```js
	if (method !== 'DELETE') {
		this.registry
			.metrics()
			.then(metrics => {
				req.write(metrics);
				req.end();
			})
			.catch(err => {
				callback(err);
			});
	} else {
		req.end();
	}
```